# Final Project: Viz For Self

Author: Shreya Udhani

The dataset we have choose for our final Project is "Pokemon with stats and types". 

# Let us explore the data

In [6]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv ('https://query.data.world/s/zpfwdyw2eisrcwdwnb44lcr5zb6vvs') 
print(df) 

       #                       Name    Type 1  Type 2  Total   HP  Attack  \
0      1                  Bulbasaur     Grass  Poison    318   45      49   
1      2                    Ivysaur     Grass  Poison    405   60      62   
2      3                   Venusaur     Grass  Poison    525   80      82   
3      3      VenusaurMega Venusaur     Grass  Poison    625   80     100   
4      4                 Charmander      Fire     NaN    309   39      52   
5      5                 Charmeleon      Fire     NaN    405   58      64   
6      6                  Charizard      Fire  Flying    534   78      84   
7      6  CharizardMega Charizard X      Fire  Dragon    634   78     130   
8      6  CharizardMega Charizard Y      Fire  Flying    634   78     104   
9      7                   Squirtle     Water     NaN    314   44      48   
10     8                  Wartortle     Water     NaN    405   59      63   
11     9                  Blastoise     Water     NaN    530   79      83   

The data can be downloaded through the link "https://query.data.world/s/zpfwdyw2eisrcwdwnb44lcr5zb6vvs". The contributor of the dataset is <b> Data Society </b>. The data is <b> Open </b> for use therefore no license is required to use the dataset.

Lets find out more about the data

We should start with basic and find out the <i> columns </i> in the dataset.

In [7]:
list(df)

['#',
 'Name',
 'Type 1',
 'Type 2',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed',
 'Generation',
 'Legendary']

There are 13 columns. Including index, Name etc. We would discuss breifly what these columns represent. 
<ol>
    <li> <b> #, Name: </b> The first 2 columns represent the index and the name of the Pokemon. </li>
    <li> <b> Type 1, Type 2</b>: Pokemons can be categorised based on their types like Grass, Poison, Flying etc. The <i> Type 1 <i> attribute tells us about the primary type of the Pokemon. In cases of some Pokemons they have a dual nature and hence they have more than one type. This is represented using <i> Type 2 </i> </li>
        <li><b> Total</b>: Total is the sum of all stats.</li>
        
<li><b>HP</b>: hit points, or health, defines how much damage a pokemon can withstand before fainting</li>

<li><b>Attack</b>: the base modifier for normal attacks (eg. Scratch, Punch)</li>

<li><b>Defense</b>: the base damage resistance against normal attacks</li>

<li><b>SP Atk</b>: special attack, the base modifier for special attacks (e.g. fire blast, bubble beam)</li>

<li><b>SP Def</b>: the base damage resistance against special attacks</li>

<li><b>Speed</b>: determines which pokemon attacks first each round</li>
    
 </ol>
 <i> Source: Data Society </i>

Number of rows and columns :

In [8]:
df.shape

(800, 13)

There are 800 rows and 13 columns. 

# Some Interesting facts

I was interested in exploring how many unique attributes there are in <i> Type 1</i>, so I found out this: 

In [9]:
df1= df['Type 1'].unique()
print(df1)

['Grass' 'Fire' 'Water' 'Bug' 'Normal' 'Poison' 'Electric' 'Ground' 'Fairy'
 'Fighting' 'Psychic' 'Rock' 'Ghost' 'Ice' 'Dragon' 'Dark' 'Steel' 'Flying']


<i> There are 18 different primary types to classify the Pokemons and the number of Pokemons classified in them are: </i>

In [13]:
pd.value_counts(df['Type 1'].values, sort=False)

Fairy        17
Ground       32
Steel        27
Grass        70
Electric     44
Psychic      57
Flying        4
Water       112
Ice          24
Poison       28
Normal       98
Fire         52
Bug          69
Fighting     27
Dragon       32
Dark         31
Ghost        32
Rock         44
dtype: int64

Total is the sum of all stats. Based on the <i> Total </i> stats, I found out what is the maximum and minimum stats for a pokemon. 

In [11]:
a = df['Total'].max()
b = df['Total'].min()
print( "The Maximum is", a )
print( "The Minimum is", b )

The Maximum is 780
The Minimum is 180


Another interesting fact was to find out how many pokemons just have <i> Type 1 </i> attribute. 

In [12]:
print("The number Of pokemons which are just Type 1 are", df['Type 2'].isnull().sum())

The number Of pokemons which are just Type 1 are 386


# Information About the data types

We can see the information about the data types and objects through:

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type 1        800 non-null object
Type 2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp. Atk       800 non-null int64
Sp. Def       800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


<b> Size Of the files </b> 

In [15]:
df.memory_usage(index= True).sum()

77680

The Memory usage is 77680 bytes. We can also see the data usage by every column: 

In [16]:
df.memory_usage(index= True)

Index           80
#             6400
Name          6400
Type 1        6400
Type 2        6400
Total         6400
HP            6400
Attack        6400
Defense       6400
Sp. Atk       6400
Sp. Def       6400
Speed         6400
Generation    6400
Legendary      800
dtype: int64